pkg/util/authtypes/authtypes.go

```diff
@@ -0,0 +1,91 @@
+package authtypes
+
+import (
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// Options contains a set of values that can be used for more fine-grained configuration of authentication.
+type Options struct {
+	// AuthoritativeSet indicates whether the agents will remove users not defined in the AutomationConfig.
+	AuthoritativeSet bool
+
+	// KeyFile is the path on disk to the keyfile that will be used for the deployment.
+	KeyFile string
+
+	// AuthMechanisms is a list of valid authentication mechanisms deployments/agents can use.
+	AuthMechanisms []string
+
+	// AgentName is username that the Automation Agent will have.
+	AgentName string
+
+	// AutoAuthMechanism is the desired authentication mechanism that the agents will use.
+	AutoAuthMechanism string
+}
+
+// Role is a struct which will map to automationconfig.Role.
+type Role struct {
+	// Name is the name of the role.
+	Name string
+
+	// Database is the database this role applies to.
+	Database string
+}
+
+// User is a struct which holds all the values required to create an AutomationConfig user
+// and references to the credentials for the specific user.
+type User struct {
+	// Username is the username of the user.
+	Username string
+
+	// Database is the database this user will be created in.
+	Database string
+
+	// Roles is a slice of roles that this user should have.
+	Roles []Role
+
+	// PasswordSecretKey is the key which maps to the value of the user's password.
+	PasswordSecretKey string
+
+	// PasswordSecretName is the name of the secret which stores this user's password.
+	PasswordSecretName string
+
+	// ScramCredentialsSecretName returns the name of the secret which stores the generated credentials
+	// for this user. These credentials will be generated if they do not exist, or used if they do.
+	// Note: there will be one secret with credentials per user created.
+	ScramCredentialsSecretName string
+
+	// ConnectionStringSecretName is the name of the secret object created by the operator
+	// which exposes the connection strings for the user.
+	// Note: there will be one secret with connection strings per user created.
+	ConnectionStringSecretName string
+
+	// ConnectionStringOptions contains connection string options for this user
+	// These options will be appended at the end of the connection string and
+	// will override any existing options from the resources.
+	ConnectionStringOptions map[string]interface{}
+}
+
+// Configurable is an interface which any resource which can configure ScramSha authentication should implement.
+type Configurable interface {
```

Comment:
I know that this has been here previously, so no need to take action, but I am wondering whether the interface gives us a lot here. Since the implementation is mdb and a mock, which afaict from first glance can be easily exchanged.